### PR TITLE
Kotlinc: catch FNFE caused by problematic destination.

### PR DESCRIPTION
**Pull request by scaventz**
This PR is related to [KT-14772](https://youtrack.jetbrains.com/issue/KT-14772) and [KT-18184](https://youtrack.jetbrains.com/issue/KT-18184)

The ```FileNotFoundException``` can be caused by:

1. Destination is not a jar, and clashing with an existing file, for example, the destination is ```a/b/c```, and the ```a``` or ```b``` or ```c```is an existing file;
2. Destination is not a jar, and permission denied, e.g: 
   - the destination is ```a/b/c/Test.class```, and lack of write permission of writting to ```c``` (**Since Windows does not support modifying the directory to be read-only, I can't provide an unit test for this situation**)
   - the destination is ```a/b/c/Test.class```, and ```Test.class``` already exists and cannot be written.
3. Destination is a jar, and output directory does not exist, we should try to create output directory for it.
4. Destination is a jar, and output directory does not exist, and failed to created output directory due to lack of permission or clash with existing file. (**Since Windows does not support modifying the directory to be read-only, I can't provide an unit test for the no permission of directory situation**)
5. Destination is a jar, and output directory exist, and permission denied to write jar. (**Since Windows does not support modifying the directory to be read-only, I can't provide an unit test for this situation**)

For these situations, I propose to deal with them as following:

cause of FNFE | message
-- | --
destination is not a jar, and clashing with an existing file | directory not found: $outputDir.
destination is not a jar, and permission denied, and output directory does not exist | directory not found: $outputDir
destination is not a jar, and permission denied, and output directory exists | error while writing $output (Permission denied)
destination is a jar, and no matter what causes the FileNotFoundException | "Invalid jar path: " + jarPath

### CODE REVIEW

**Comment by scaventz** [comment]
Converted to draft to resolve conflicts.

File: compiler/tests/org/jetbrains/kotlin/integration/CompilerSmokeTest.java [hunk 1]
```diff
@@ -201,4 +201,50 @@ public void testNoReflect() throws Exception {
...
+        runCompiler("test.compile", "test.kt", "-d", outputDir);
+    }
+
+    // related to KT-18184, destination is not a jar, and permission denied
+    public void testDestinationDirNoPermission() throws Exception {
+        String outputDir = tmpdir.getAbsolutePath() + File.separator + "noPermissionDir";
+        File file = new File(outputDir, "Test.class");
+        file.getParentFile().mkdirs();
+        file.createNewFile();
+        // file.getParentFile().setReadOnly(); // won't work on Windows
```
Comment (udalov) [inline compiler/tests/org/jetbrains/kotlin/integration/CompilerSmokeTest.java:219]: I guess we could wrap this into `if (SystemInfo.isWindows)` to make sure we test the original issue at least on Linux, but the current way is also OK

**Comment by udalov** [comment]
Sorry for the delay. The changes look good, I've merged them. The only change I did was move `NoPermissionException` back to `cli` because it's only used there.

Thanks for the contribution! Special thanks for paying attention for all the cases and testing them properly.

**Comment by udalov** [comment]
@scaventz Looks like it leads to NPE if the output is a jar with a relative path in the current directory, because `jarPath.getParentFile()` returns `null`:
```
kotlinc 1.kt -d out.jar
```
```
exception: java.lang.NullPointerException
	at org.jetbrains.kotlin.cli.jvm.compiler.CompileEnvironmentUtil.writeToJar(CompileEnvironmentUtil.java:119)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.writeOutput(KotlinToJVMBytecodeCompiler.kt:91)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.writeOutputs(KotlinToJVMBytecodeCompiler.kt:228)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli(KotlinToJVMBytecodeCompiler.kt:215)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli$default(KotlinToJVMBytecodeCompiler.kt:155)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:169)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:52)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:90)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:44)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:98)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:76)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:45)
	at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMainNoExit(CLITool.kt:227)
	at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMainNoExit$default(CLITool.kt:222)
	at org.jetbrains.kotlin.cli.common.CLITool$Companion.doMain(CLITool.kt:214)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler$Companion.main(K2JVMCompiler.kt:271)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.main(K2JVMCompiler.kt)
```

**Comment by scaventz** [comment]
@udalov Thank you for spotting this bug, I've opened another pull request to fix it. I should also provide a new test on it, but currently it seems tests all work with temp directory and cannot simulate this situation. Thinking about it..

**Pull request merged**